### ISSUE TITLE
Cancellation tokens passed to constructors

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn-analyzers#4467]
**Diagnostic ID**: [CA1068](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1068)

Would it make sense for this rule to ignore constructors on the premise that constructors are more likely to be assigning the cancellation token to a field than using it?

**Comment by Evangelink** [comment]
Out of curiosity, could you give some examples of such construction? It seems a little weird to have the token in the ctor instead of on all long running methods.

Putting aside my previous comment and given that a constructor is not supposed to be a long running, cancellable, action, I would say yes it makes sense. Besides, the documentations states:
> However, the cancellation token itself is not usually relevant to the core functionality of a majority of these methods.
In this case, the cancellation token is actually part of the core functionality.

@mavasani What's your opinion here?

Note that we could consider adding the `excluded_symbol_names` which supports `.ctor` exclusion which would solve the issue. I think it's still worth considering what makes sense from a design point of view rather than directly jumping to the implementation of this user option.

**Comment by mavasani** [comment]
I think we likely want to retain existing default behavior for the analyzer, but can potentially offer an option to configure not flagging constructors as @Evangelink mentioned. Note that lot of code bases would still likely prefer token to be ctor's last parameter as a style preference, so changing the default behavior would be a subtle behavior change.

**Comment by jnm2** [comment]
Here's my only example:

```cs
private sealed class Barrier
{
    private readonly System.Threading.Barrier barrier;
    private readonly CancellationToken cancellationToken;
    private readonly bool[] wasWaitingWhenCanceled;
    private readonly int index;


    public Barrier(System.Threading.Barrier barrier, CancellationToken cancellationToken, bool[] wasWaitingWhenCanceled, int index)
    {
        this.barrier = barrier;
        this.cancellationToken = cancellationToken;
        this.wasWaitingWhenCanceled = wasWaitingWhenCanceled;
        this.index = index;
    }

    public void SignalAndWait()
    {
        Volatile.Write(ref wasWaitingWhenCanceled[index], true);
        barrier.SignalAndWait(cancellationToken);
        Volatile.Write(ref wasWaitingWhenCanceled[index], false);
    }
}
```
